[Flaky] /spec/system/admin/bulk_order_management_spec

### What we should change and why (this is tech debt)
<!-- The use of X changed and part y is now redundant. We can remove it and... -->
We already have #9899 for this spec but it seems like a different test is failing. Mentioning here in case they are still related.

```
rspec ./spec/system/admin/bulk_order_management_spec.rb:204 # 
  As an Administrator
  I want to be able to manage orders in bulk
 listing orders searching by email
I, [2023-12-26T15:11:36.309118 #2866]  INFO -- : [knapsack_pro] To retry all the tests assigned to this CI node, please run the following command on your machine:
I, [2023-12-26T15:11:36.309338 #2866]  INFO -- : [knapsack_pro] bundle exec rspec --format progress --default-path spec "spec/system/admin/bulk_order_management_spec.rb"
Stopped processing SimpleCov as a previous error not related to SimpleCov has been detected
Coverage report generated for RSpec to /home/runner/work/openfoodnetwork/openfoodnetwork/coverage. 6584 / 25329 LOC (25.99%) covered.
Error: Process completed with exit code 1.
```


### Context
<!-- This came up in PR #3451#issuecomment-461616636. -->
https://github.com/openfoodfoundation/openfoodnetwork/actions/runs/7330546276/job/19961836397?pr=11982


### Impact and timeline
<!-- Example: This simplifies the upgrade to Spree 3 and should be done after upgrading to Spree 2. -->

--- a/spec/system/admin/request_scoped_login_spec.rb
+++ b/spec/system/admin/request_scoped_login_spec.rb
@@ -8,16 +8,38 @@ require 'net/http'
 # firing a "stray" request straight at the Capybara test server — as if it had leaked from a
 # previous example — so that it reaches Warden *after* `login_as` but *before* our `visit`.
 # Without request scoping that stray request consumes the queued login and we land
-# unauthenticated; with it, the login stays bound to our own (header-marked) request.
+# unauthenticated; with it, the login stays bound to our own (token-marked) request.
 RSpec.describe "Request-scoped Warden test login" do
   include AuthenticationHelper
+
+  def fire_stray_request(headers = {})
+    server = Capybara.current_session.server
+    uri = URI("http://#{server.host}:#{server.port}/admin")
+    Net::HTTP.start(uri.host, uri.port) do |http|
+      request = Net::HTTP::Get.new(uri)
+      headers.each { |name, value| request[name] = value }
+      http.request(request)
+    end
+  end
 
   it "keeps the login bound to our request even if a stray request reaches Warden first" do
     login_as_admin
 
-    # A request with no marker header, exactly like one leaking from the previous example.
-    server = Capybara.current_session.server
-    Net::HTTP.get_response(URI("http://#{server.host}:#{server.port}/admin"))
+    # A request with no marker header, exactly like one leaking from an example that never
+    # logged in.
+    fire_stray_request
+
+    visit spree.edit_admin_tax_settings_path
+
+    expect(page).to have_css("body.admin")
+  end
+
+  it "is not stolen by a stray request carrying a previous example's marker token" do
+    login_as_admin
+
+    # A leak from a *previous logged-in* example carries a marker header, but with that
+    # example's own (now stale) token value — never our fresh one.
+    fire_stray_request(RequestScopedLogin::HEADER => "stale-token-from-previous-example")
 
     visit spree.edit_admin_tax_settings_path
 

--- a/spec/system/support/request_scoped_login.rb
+++ b/spec/system/support/request_scoped_login.rb
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require "securerandom"
+
 # Binds Warden's test login to the specific request the example makes, instead of
 # "the next request to reach Warden".
 #
@@ -11,42 +13,54 @@
 # login against its own throw-away session, and leave the real page unauthenticated and
 # redirected to the public login screen. See teamcapybara/capybara#702 and #1692.
 #
-# Fix: tag the example's own request with a marker HTTP header and only apply the queued
-# login to a request carrying that header. A leaked request can't carry it — Chrome cannot
-# retro-add a header (`Network.setExtraHTTPHeaders`) to an already-dispatched request — so
-# it can no longer steal the login. One-shot semantics are preserved: the login is applied
-# exactly once, then the session cookie carries it, so logout / forbidden-redirect still work.
+# Fix: tag the example's own request with a marker HTTP header whose value is a fresh,
+# per-`login_as` token, and apply the queued login only to a request carrying THAT token.
+# A leaked request cannot carry it:
+# - a request from a previous example carries that example's (now stale) token, and
+# - Chrome cannot retro-add a header (`Network.setExtraHTTPHeaders`) to an already-dispatched
+#   request,
+# so a leak can never match the current example's token and can no longer steal the login.
+# Using a constant marker value is not enough: every logged-in example sets the same header,
+# so a leak from the previous logged-in example carries it and still races us (this is why
+# the earlier constant-value version stayed flaky).
 #
-# Applied to every system spec; the Warden hook is header-gated, so examples that never call
-# `login_as` are unaffected (no header is set, the holder stays empty, the hook no-ops).
+# One-shot semantics are preserved: the login is keyed by token and deleted on first use, then
+# the session cookie carries it, so logout / forbidden-redirect still work.
+#
+# Applied to every system spec; the Warden hook is token-gated, so examples that never call
+# `login_as` are unaffected (no header is set, so no request carries a known token and the
+# hook no-ops).
 module RequestScopedLogin
   HEADER = "X-Warden-Test-Login"
   ENV_KEY = "HTTP_X_WARDEN_TEST_LOGIN"
 
-  # One-shot holder, mirrors Warden._on_next_request but drained only by the marked request.
+  # One-shot logins keyed by token; each is drained only by a request carrying its token.
   def self.pending
-    @pending ||= []
+    @pending ||= {}
   end
 
   def self.install_warden_hook!
     return if @installed
 
     Warden::Manager.on_request do |proxy|
-      next unless proxy.env[ENV_KEY]
+      token = proxy.env[ENV_KEY]
+      next unless token
 
-      while (entry = pending.shift)
-        user, opts = entry
-        proxy.set_user(user, opts)
-      end
+      entry = pending.delete(token)
+      next unless entry
+
+      user, opts = entry
+      proxy.set_user(user, opts)
     end
     @installed = true
   end
 
-  # Overrides Warden::Test::Helpers#login_as for tagged examples.
+  # Overrides Warden::Test::Helpers#login_as for system specs.
   def login_as(user, opts = {})
     opts = { event: :authentication }.merge(opts)
-    RequestScopedLogin.pending.replace([[user, opts]])
-    page.driver.add_header(RequestScopedLogin::HEADER, "1")
+    token = SecureRandom.hex(16)
+    RequestScopedLogin.pending[token] = [user, opts]
+    page.driver.add_header(RequestScopedLogin::HEADER, token)
   end
 end
 

